src/induction/base.ml

```diff
@@ -193,18 +221,40 @@ let rec next (input_sys, aparam, trans, solver, k, unknowns) =
     (* Asserting implications if k > 0. *)
     if Numeral.(k > zero) then
       nu_unknowns
+      (* If we're skipping steps, we don't need to assert that the property
+         held in the previous step *)
+      |> List.filter (fun (name, _) -> match TransSys.get_prop_kind trans name with
+        | Reachable Some (From b)
+        | Reachable Some (At b) -> b < k_int
+        | Reachable Some (FromWithin (b, _))-> b < k_int
+        | _ -> true
+      ) 
       |> List.map (fun (_, term) -> Term.bump_state Numeral.(k-one) term)
       |> Term.mk_and |> SMTSolver.assert_term solver ;
 
+    (* (Temporarily) filter out properties with a lower bound smaller than
+       current k. *)
+    let out_of_bounds, nu_unknowns = nu_unknowns |> List.partition (
+      fun (name,_) -> match TransSys.get_prop_kind trans name with
+      | Reachable Some (From b)
+      | Reachable Some (At b) -> b > k_int
+      | Reachable Some (FromWithin (b, _))-> b > k_int
+      | _ -> false
+    ) in
     (* Filtering properties which are not known to be k-true at this step. *)
     let unknowns_at_k, k_true =
       nu_unknowns |> List.partition (
-        fun (name,_) -> match TransSys.get_prop_status trans name with
-        | Property.PropKTrue k -> k_int > k
+        fun (name,_) -> match (TransSys.get_prop_status trans name, TransSys.get_prop_kind trans name) with
+        | (Property.PropKTrue k, Invariant) 
+        | (Property.PropKTrue k, Reachable None) -> k_int > k
+        | (Property.PropKTrue k, Reachable Some (From b)) -> k_int > k && b <= k_int
+        | (Property.PropKTrue k, Reachable Some (Within b)) -> k_int > k && b >= k_int
+        | (Property.PropKTrue k, Reachable Some (At b)) -> k_int > k && b = k_int
+        | (Property.PropKTrue k, Reachable Some (FromWithin (b1, b2))) -> k_int > k && b1 <= k_int && b2 >= k_int
         | _ -> true
       )
     in
-
+    let step = ref k in
     let unfalsifiable =
```

Comment:
Please avoid code with side effects if possible. In this case, compute two values `let unfalsifiable, k =`. Then replace `if not invs then skip_steps_next trans solver step k;` with `k = if not invs then skip_steps_next trans solver k k else k` and `k_true @ unfalsifiable` with `k_true @ unfalsifiable, k`. It is fine to use a local reference within ` skip_steps_next` to implement the while loop. The important point is that any use of a (stateful) variable is local to the function that use it and is not shared by the caller.